fs/dafsserver/dafsserver.cpp

```diff
@@ -1394,21 +1452,552 @@ class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
+    {
+        maxRowSize = config.getPropInt64("maxRowSize", defaultMaxCsvRowSize) * 1024 * 1024;
+        preserveWhitespace = config.getPropBool("preserveWhitespace");
+
+        if (!config.getProp("csvQuote", csvQuote))
+        {
+            if (!fileDesc->queryProperties().getProp("@csvQuote", csvQuote))
+                csvQuote.append("\"");
+        }
+        if (!config.getProp("csvSeparate", csvSeparate))
+        {
+            if (!fileDesc->queryProperties().getProp("@csvSeparate", csvSeparate))
+                csvSeparate.append("\\,");
+        }
+        if (!config.getProp("csvTerminate", csvTerminate))
+        {
+            if (!fileDesc->queryProperties().getProp("@csvTerminate", csvTerminate))
+                csvTerminate.append("\\n,\\r\\n");
+        }
+        if (!config.getProp("csvEscape", csvEscape))
+            fileDesc->queryProperties().getProp("@csvEscape", csvEscape);
+
+        headerLines = config.getPropInt64("headerLines"); // really this should be a published attribute too
+    }
+    virtual StringBuffer &getInfoStr(StringBuffer &out) const override
+    {
+        return out.appendf("csvread[%s]", fileName.get());
+    }
+// IRemoteReadActivity impl.
+    virtual const void *nextRow(MemoryBufferBuilder &outBuilder, size32_t &retSz) override
+    {
+        if (eofSeen)
+        {
+            retSz = 0;
+            return nullptr;
+        }
+        checkOpen();
+        while (!eofSeen && (processed < chooseN))
+        {
+            size32_t lineLength = csvSplitter.splitLine(inputStream, maxRowSize);
+            if (!lineLength)
+                break;
+
+            retSz = translator->translate(outBuilder, *this, *fieldFetcher);
+
+            if (retSz) // could it ever be 0 (skip) ??
+            {
+                const void *ret = outBuilder.getSelf();
+                if (fieldFilterMatch(ret))
+                {
+                    outBuilder.finishRow(retSz);
+                    ++processed;
+                    inputStream->skip(lineLength);
+                    return ret;
+                }
+                else
+                    outBuilder.removeBytes(retSz);
+            }
+            inputStream->skip(lineLength);
+        }
+        eofSeen = true;
+        close();
+        retSz = 0;
+        return nullptr;
+    }
 };
 
 
-IRemoteReadActivity *createRemoteDiskRead(IPropertyTree &actNode)
+class CRemoteMarkupReadActivity : public CRemoteExternalFormatReadActivity, implements IXMLSelect
 {
-    return new CRemoteDiskReadActivity(actNode);
-}
+    typedef CRemoteExternalFormatReadActivity PARENT;
+
+    ThorActivityKind kind;
+    IXmlToRowTransformer *xmlTransformer;
+    Linked<IColumnProvider> lastMatch;
+    Owned<IXMLParse> xmlParser;
+
+    StringBuffer xpath;
+    bool noRoot = false;
+    bool useXmlContents = false;
+
+    // JCSMORE - it would be good if these were cached/reused (can I assume anything using fetcher is single threaded?)
```

Comment:
Once the interfaces are refactored I think this will be clearer (and can be reused)